core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -465,6 +465,79 @@ void cuda_prefetch_pointer(const Cuda &space, const void *ptr, size_t bytes,
 }
 
 }  // namespace Impl
+
+namespace {
+
+template <typename RequestedMemorySpace>
+void check_memory_space(cudaMemoryType deduced_memory_type) {
+  switch (deduced_memory_type) {
+    case cudaMemoryTypeHost:
+      if (!Kokkos::SpaceAccessibility<RequestedMemorySpace,
+                                      Kokkos::CudaHostPinnedSpace>::assignable)
+        Kokkos::abort(
+            ("Detected CudaHostPinnedSpace but requested incompatible " +
+             std::string(RequestedMemorySpace::name()))
+                .c_str());
+      return;
+    case cudaMemoryTypeDevice:
+      if (!Kokkos::SpaceAccessibility<RequestedMemorySpace,
+                                      Kokkos::CudaSpace>::assignable)
+        Kokkos::abort(("Detected CudaSpace but requested incompatible " +
+                       std::string(RequestedMemorySpace::name()))
+                          .c_str());
+      return;
+    case cudaMemoryTypeManaged:
+      if (!Kokkos::SpaceAccessibility<RequestedMemorySpace,
+                                      Kokkos::CudaUVMSpace>::assignable)
+        Kokkos::abort(("Detected CudaUVMSpace but requested incompatible " +
+                       std::string(RequestedMemorySpace::name()))
+                          .c_str());
+      return;
+    case cudaMemoryTypeUnregistered:
```

Comment:
So a pointer allocated using `malloc` on a machine with CUDA 12 and a kernel that supports HMM would hit the `Managed` path?

Reply:
I would expect the runtime to say `Unregistered` but I haven't checked.

Reply:
If I understand https://developer.nvidia.com/blog/simplifying-gpu-application-development-with-heterogeneous-memory-management/#system_allocated_memory_and_cuda_allocators correctly, basically all memory allocated with traditional stuff like `mmap` or `malloc` should be accessible on the GPU and CPU if HMM is available and the driver supports it (independently of the GPU) ... I guess we would need to test if it results in `Unregistered` or `Managed`.

Reply:
See https://github.com/kokkos/kokkos/issues/7597#issuecomment-2541733930 for Grace Hopper. In short, the runtime detection only cares about the allocation mechanism. `malloc` is mapped to `Unregistered`.

Reply:
Also note that we don't change any of the `SpaceAccessibility<...>::assignable` values for MI300A or GH200.

Reply:
Thanks for the tests, @masterleinad. I guess my point is: should we treat something that the runtime reports `Unregistered` as `Managed`, if it is malloced on coda 12.2 with HMM? because it can migrate similar to UVM.

Reply:
In my opinion, all considerations should go into `SpaceAccessibility`/`MemorySpaceAccess`. We can certainly say that `HostSpace` is assignable to `CudaUVMSpace` but I wouldn't mess with the detection here.

Reply:
yeah, I am all for deciding in the `SpaceAccessibility`. I just think we should make a plan how to deal with the somewhat new memory behavior in general. That was the intention of https://github.com/kokkos/kokkos/issues/7607